lib/eco/autolboxdetector.py

```diff
@@ -42,24 +42,98 @@ def pop_lookahead(self, node):
+            # if the user had to revert an automatically inserted box. However,
+            # we might want to consider showing suggestions if the suggested box
+            # differs from the one the user reverted
+            return False # don't use this node for autoboxes anymore
+        if valid:
+            errornode.autobox = valid
+            return True
+        else:
+            errornode.autobox = None
+            return False
+
+    def detect_lbox_h2(self, errornode):
+        valid = []
+        ws = ["<ws>", "<return>"]
+        searched = set()
+        pv = self.op.prev_version
+        for sub in self.langs:
+            lbox = MagicTerminal("<{}>".format(sub))
+            parent = errornode.parent
+            while parent is not None:
+                if parent.get_attr("parent", pv) is None: # Root
+                    # If we've reached the root, try inserting the box after
+                    # BOS, i.e. the beginning of the file
+                    left = parent.get_attr("children", pv)[0] # bos
+                else:
+                    left = parent.get_attr("left", pv)
+                if left:
+                    state = left.state
+                    element = self.op.syntaxtable.lookup(state, lbox)
+                    if type(element) in [Reduce, Shift]:
+                        term = self.find_terminal(left, pv)
+                        if type(term) is EOS:
+                            continue
+                        while term and term.lookup in ws:
+                            # skip whitespace
+                            term = term.next_term
+                        if term and term not in searched:
+                            r = self.langs[sub]
+                            r.mode_limit_tokens_new = self.mode_limit_tokens_new
+                            result = r.parse(term)
+                            if r.possible_ends:
+                                for e in r.possible_ends:
+                                    if e.lookup in ws:
+                                        continue
+                                    if (self.contains_errornode(term, e, errornode) \
+                                            and self.parse_after_lbox_h2(lbox, e, parent)):
+                                                valid.append((term, e, sub))
+                                                searched.add(term)
+                parent = parent.get_attr("parent", pv)
+        return valid
+
+    def parse_after_lbox_h2(self, lbox, end, parent):
+        root = parent.get_root()
+        p, l = lang_dict[root.name].load()
+        ir = IncrementalRecognizer(p.syntaxtable, l.lexer, root.name, None)
+        if root is not parent:
+            # if the parent is already the root we don't need to preparse
+            # anything
+            ir.preparse(root, parent)
+        # try parsing lbox + one more non-ws terminal
+        if ir.parse_single(TextNode(lbox)):
```

Comment:
You could simplify these lines to `return ir.parse_single(TextNode(lbox)) and ir.parse_after(end.next_term)`.